teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,997 @@
+
+However, since the `StepAction` are defined in isolation and are referenced in `Tasks`, it is challenging for the `Task` authors to pass `Results` from one `Step` to another without Tekton providing some API. Therefore, we should enable referencing of `Results` produced by `Steps` via `Parameters` of the subsequent `Steps`. 
+
+`Task` authors should be able to extract `Results` from previous `Steps` (i.e. written to  `$(step.results.resultName.path)`) using `$(steps.stepName.results.resultName)`. `Results` written to `$(results.resultName.path)` are not referenceable.
+
+<table>
+<tr>
+<td>
+
+```yaml
+# This is how a standalone step author 
+# would declare the results that it is producing.
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  results:
+    - name: foo
+  image: ...
+  script: |
+    echo “hi” >> $(step.results.foo.path)
+```
+</td>
+<td rowspan=2>
+
+```yaml
+# This is how a task author would 
+# use the results from one step 
+# into subsequent steps.
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  results:
+    - name: res1
+      value: $(steps.step1.results.foo)
+  steps:
+    - name: step1
+      stepRef:
+        name: myStep
+    - name: step2
+      stepRef:
+        name: myOtherStep
+      params:
+        - name: p1
+          value:  $(steps.step1.results.foo)
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myOtherStep
+spec:
+  params:
+    - name: p1
+  image: ...
+  script: |
+    echo $(params.p1)
+
+```
+</td>
+</tr>
+</table>
+
+#### Workspaces
+`Workspace` is an orchestration concept because it’s bound to a `Task`. All the `Steps` in a `Task` can access the Workspaces bound to the `Task`. Occasionally, a `Task` author may want to provide a `Step` exclusive access to some Workspace which they can do via the `Workspaces` field in `Steps`.
+
+The actionable components of a `Step` (i.e. `commands`, `args` and `script`) only require a “path” to a `Workspace` which they refer to via `$(workspaces.someworkspace.path)`. This variable could also be passed in as a string `Parameter` instead. 
+
+When defining an inline `Step`, the `Task` authors have `Workspace` information in context so they can refer to `Workspace` paths directly. However, when referencing a `StepAction`, the `Task` authors should pass paths via `Parameters` because there are no `Workspaces` in `StepAction`.  A `StepAction` should expect to “just work” if all the context is set up appropriately by the `Task` and the correct inputs are passed to it.
+
+This means that if users are migrating inlined `Steps` to `StepAction`, the rules are slightly different so they need to convert an `Workspace` path into a `Parameter` before using it.
```

Comment:
How does it "work" with "Isolated workspaces" ? https://tekton.dev/docs/pipelines/workspaces/#isolating-workspaces-to-specific-steps-or-sidecars

Reply:
I think in the same way. If I understand correctly, `isolatedWorkspaces` is just telling the controller to only bind the workspace to a specific container. When used in the `actionable` part of the step, a Task author is still doing something like `$(workspaces.workspaceName.path)` or something which could just as easily be passed as a parameter.

Reply:
Good point 👼🏼